sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventBatch.cs

```diff
@@ -183,12 +174,59 @@ public override void Clear()
         ///
         public override IReadOnlyCollection<T> AsReadOnlyCollection<T>()
         {
-            if (typeof(T) != typeof(EventData))
+            if (typeof(T) != typeof(AmqpMessage))
             {
                 throw new FormatException(string.Format(CultureInfo.CurrentCulture, Resources.UnsupportedTransportEventType, typeof(T).Name));
             }
 
-            return BatchEvents as IReadOnlyCollection<T>;
+            return _batchMessages as IReadOnlyCollection<T>;
+        }
+
+        /// <summary>
+        ///   Assigns message sequence numbers and publisher metadata to the batch in
+        ///   order to prepare it to be sent when certain features, such as idempotent retries,
+        ///   are active.
+        /// </summary>
+        ///
+        /// <param name="lastSequenceNumber">The sequence number assigned to the event that was most recently published to the associated partition successfully.</param>
+        /// <param name="producerGroupId">The identifier of the producer group for which publishing is being performed.</param>
+        /// <param name="ownerLevel">TThe owner level for which publishing is being performed.</param>
+        ///
+        /// <returns>The last sequence number applied to the batch.</returns>
+        ///
+        public override int ApplyBatchSequencing(int lastSequenceNumber,
+                                                 long? producerGroupId,
+                                                 short? ownerLevel)
+        {
+            if (_batchMessages.Count == 0)
+            {
+                return lastSequenceNumber;
+            }
+
+            _startingSequenceNumber = NextSequence(lastSequenceNumber);
```

Comment:
I wanted to avoid embedding an `if` statement in the `foreach` loop, so I'm paying the cost of the unchecked increment/rollover twice rather than the conditional for each tick.